### PR TITLE
Feature: Add `Raft::external_state_machine_request()` to run a function inside state machine

### PR DESCRIPTION
**Pull request by drmingdrmer**
##### Feature: Add `Raft::external_state_machine_request()` to run a function inside state machine

Add `Raft::external_sm_request()` in a fire-and-forget manner,
and `Raft::with_state_machine()` blocks waiting on the response.

Other changes: move OptionalSerde, OptionalSync, OptionalSync from
`openraft::` to `openraft::base::`

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1206)

### CODE REVIEW

**Review by schreter — APPROVED** [review]
> Reviewed 12 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1206)* status: all files reviewed, 1 unresolved discussion (waiting on @drmingdrmer)

___
*[`openraft/src/core/raft_core.rs` line 1220 at r1](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2jty7_9L010i5ZlX7k:-O2jty7_9L010i5ZlX7l:bm0d7po) ([raw file](https://github.com/datafuselabs/openraft/blob/47fcca8793b17716680b0434a8f736c5488f0172/openraft/src/core/raft_core.rs#L1220)):*
> ```XML
>                         let res = self.sm_handle.send(sm_cmd);
>                         if let Err(e) = res {
>                             tracing::error!(error = display(e), "error sending sm::Command to sm::Worker");
> ```

BTW, I noticed you use `display()` at many places. You know there are also shortcuts in `tracing`, like `%x` for `display()`?

OTOH, `display()` is more speaking than some random `%` character...



_Suggestion:_
```XML
tracing::error!(error = %e, "error sending sm::Command to sm::Worker");
```
___
*[`openraft/src/raft/mod.rs` line 831 at r1](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2juuOjDKREr7_T89kj:-O2juuOjDKREr7_T89kk:b622oiv) ([raw file](https://github.com/datafuselabs/openraft/blob/47fcca8793b17716680b0434a8f736c5488f0172/openraft/src/raft/mod.rs#L831)):*
> ```XML
>     ///
>     /// The request functor will be called with a mutable reference to the state machine
>     /// and.
> ```

`and` what? (docs)



_Code quote:_
```XML
    /// The request functor will be called with a mutable reference to the state machine
    /// and.
```

**Review by drmingdrmer — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1206)* status: all files reviewed, 1 unresolved discussion (waiting on @schreter)

___
*[`openraft/src/core/raft_core.rs` line 1220 at r1](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2jty7_9L010i5ZlX7k:-O2m3Wum4WIoRGQyc6IC:b-sr3433) ([raw file](https://github.com/datafuselabs/openraft/blob/47fcca8793b17716680b0434a8f736c5488f0172/openraft/src/core/raft_core.rs#L1220)):*
<details><summary><i>Previously, schreter wrote…</i></summary><blockquote>

BTW, I noticed you use `display()` at many places. You know there are also shortcuts in `tracing`, like `%x` for `display()`?

OTOH, `display()` is more speaking than some random `%` character...
</blockquote></details>


I know that but some IDE fails to parse `%e/?e` and emit an error.
I prefer to use function call like syntax if possible. :thinking: 
___
*[`openraft/src/raft/mod.rs` line 831 at r1](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2juuOjDKREr7_T89kj:-O2m45f45N2cJ_vR0K4t:bmjgntt) ([raw file](https://github.com/datafuselabs/openraft/blob/47fcca8793b17716680b0434a8f736c5488f0172/openraft/src/raft/mod.rs#L831)):*
<details><summary><i>Previously, schreter wrote…</i></summary><blockquote>

`and` what? (docs)
</blockquote></details>

:worried:

**Review by schreter — DISMISSED** [review]
> Reviewed 8 of 9 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1206)* status: all files reviewed, 4 unresolved discussions (waiting on @drmingdrmer)

___
*[`openraft/src/base/mod.rs` line 26 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2o_0ZT-iKnvIxEpN8j:-O2o_0ZT-iKnvIxEpN8k:bbus5qy) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/base/mod.rs#L26)):*
> ```XML
>     pub type BoxAsyncOnceMut<'a, A, T = ()> = Box<dyn FnOnce(&mut A) -> BoxFuture<T> + Send + 'a>;
>     pub type BoxOnce<'a, A, T = ()> = Box<dyn FnOnce(&A) -> T + Send + 'a>;
>     pub type BoxAny = Box<dyn Any + Send>;
> ```

Another option would be to create a helper trait `Any + OptionalSend` in common code and implement it for all `T`, but this is also fine. However, see below.
___
*[`openraft/src/core/raft_core.rs` line 1220 at r1](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2jty7_9L010i5ZlX7k:-O2o_DWJ5UHEFG0Qaurh:b2woun5) ([raw file](https://github.com/datafuselabs/openraft/blob/47fcca8793b17716680b0434a8f736c5488f0172/openraft/src/core/raft_core.rs#L1220)):*
<details><summary><i>Previously, drmingdrmer (张炎泼) wrote…</i></summary><blockquote>

I know that but some IDE fails to parse `%e/?e` and emit an error.
I prefer to use function call like syntax if possible. :thinking:
</blockquote></details>

No problem, I just wanted to point it out. And yes, `display()` is more speaking.

BTW, IDEs my colleagues are using (VS Code, VIM and Emacs) all render it correctly.
___
*[`openraft/src/core/sm/worker.rs` line 146 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2oa3R939S5bECfB19_:-O2oa3R939S5bECfB19a:bhc0nrr) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/core/sm/worker.rs#L146)):*
> ```XML
>                     let _ = self.resp_tx.send(Notification::sm(res));
>                 }
>                 Command::Func { func, input_sm_type } => {
> ```

Regarding the trait above, the helper trait for `func` would of course also allow you to define and auto-implement methods, like `type_name()`, so you don't need to carry type name with you.
___
*[`openraft/src/core/sm/worker.rs` line 154 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2oay8IC1uxDS6HtRo5:-O2oay8IC1uxDS6HtRo6:b3kre) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/core/sm/worker.rs#L154)):*
> ```XML
>                     } else {
>                         tracing::warn!(
>                             "User defined SM function uses incorrect state machine type, expect: {}, got: {}",
> ```

?



_Suggestion:_
```XML
"User-defined SM function uses incorrect state machine type, expected: {}, got: {}"
```
___
*[`openraft/src/raft/mod.rs` line 800 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2oaQWC1jBfjPplXOp3:-O2oaQWC1jBfjPplXOp4:bvt3qio) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/raft/mod.rs#L800)):*
> ```XML
>     ///
>     /// If the user function fail to run, e.g., the input `SM` is different one from the one in
>     /// `RaftCore`, it returns a `&'static str` error.
> ```

And what is in the `&'static str`? I'd suggest to return a wrapper, such as `struct InvalidStateMachineType { pub actual_type: &'static str };`.
___
*[`openraft/src/raft/mod.rs` line 832 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2obBqACCbgw-vhD8vD:-O2obBqACCbgw-vhD8vE:b3kre) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/raft/mod.rs#L832)):*
> ```XML
>         let Ok(v) = recv_res else {
>             if self.inner.is_core_running().await {
>                 return Ok(Err("State machine user defined function fail to execute."));
> ```

?



_Suggestion:_
```XML
User-defined function on the state machine failed to execute
```

**Review by drmingdrmer — COMMENTED** [review]
> Reviewed 9 of 9 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1206)* status: 13 of 16 files reviewed, 2 unresolved discussions (waiting on @schreter)

___
*[`openraft/src/core/sm/worker.rs` line 146 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2oa3R939S5bECfB19_:-O2rLQEo8hE9J3vrfisc:blcvojh) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/core/sm/worker.rs#L146)):*
<details><summary><i>Previously, schreter wrote…</i></summary><blockquote>

Regarding the trait above, the helper trait for `func` would of course also allow you to define and auto-implement methods, like `type_name()`, so you don't need to carry type name with you.
</blockquote></details>

It looks like such an approach only outputs `"alloc::boxed::Box<dyn core::any::Any>"`
Is there anything I missed?

```rust
use std::any::Any;

trait Foo: Any {
    fn get_type(&self) -> &'static str;
}

impl<T: Any> Foo for T {
    fn get_type(&self) -> &'static str {
        std::any::type_name::<T>()
    }
}

fn main() {
    let a: Box<dyn Any> = Box::new(5);
    println!("{}", a.get_type());
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7058005d5cead64910f983dc1d8145d2
___
*[`openraft/src/core/sm/worker.rs` line 154 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2oay8IC1uxDS6HtRo5:-O2rARPs89ZhS44o5yst:bhgxvih) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/core/sm/worker.rs#L154)):*
<details><summary><i>Previously, schreter wrote…</i></summary><blockquote>

?
</blockquote></details>

Thanks for these minor typo fixes!

**Review by schreter — DISMISSED** [review]
> Reviewed 3 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1206)* status: all files reviewed, 2 unresolved discussions (waiting on @drmingdrmer)

___
*[`openraft/src/core/sm/worker.rs` line 146 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2oa3R939S5bECfB19_:-O2tsz5r05OyRf-SyffK:bf7txqv) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/core/sm/worker.rs#L146)):*
<details><summary><i>Previously, drmingdrmer (张炎泼) wrote…</i></summary><blockquote>

It looks like such an approach only outputs `"alloc::boxed::Box<dyn core::any::Any>"`
Is there anything I missed?

```rust
use std::any::Any;

trait Foo: Any {
    fn get_type(&self) -> &'static str;
}

impl<T: Any> Foo for T {
    fn get_type(&self) -> &'static str {
        std::any::type_name::<T>()
    }
}

fn main() {
    let a: Box<dyn Any> = Box::new(5);
    println!("{}", a.get_type());
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7058005d5cead64910f983dc1d8145d2
</blockquote></details>

Well, of course, since it's implemented for `Any`, it's also implemented for `Box`, so it won't auto-deref. I.e., you are printing the type of the `Box<dyn Any>` (in fact, it should read `Box<dyn Foo>`), not of the contained `T`.

If you'd rewrite it with `a.as_ref().get_type()` or similar (dereferencing the `Box`), you'd see the correct integer type.

Unfortunately, Rust playground doesn't work at the moment, so I can't update the code there
___
*[`openraft/src/error/invalid_sm.rs` line 15 at r5](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2ttlfU1gJl_06x5ODb:-O2ttlfU1gJl_06x5ODc:b3kre) ([raw file](https://github.com/datafuselabs/openraft/blob/8af6cf15fb72449ac092783033ca6f6c55840f95/openraft/src/error/invalid_sm.rs#L15)):*
> ```XML
>         )
>     }
> }
> ```

?



_Suggestion:_
```XML
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
#[error("User-defined function on the state machine failed to run; It may have used a different type \
         of state machine from the one in RaftCore (`{actual_type}`)")]
pub struct InvalidStateMachineType {
    pub actual_type: &'static str,
}
```

**Review by drmingdrmer — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1206)* status: 16 of 17 files reviewed, 1 unresolved discussion (waiting on @schreter)

___
*[`openraft/src/core/sm/worker.rs` line 146 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2oa3R939S5bECfB19_:-O2twlkwBWw5iVuPS0o_:b-cwsg5o) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/core/sm/worker.rs#L146)):*
<details><summary><i>Previously, schreter wrote…</i></summary><blockquote>

Well, of course, since it's implemented for `Any`, it's also implemented for `Box`, so it won't auto-deref. I.e., you are printing the type of the `Box<dyn Any>` (in fact, it should read `Box<dyn Foo>`), not of the contained `T`.

If you'd rewrite it with `a.as_ref().get_type()` or similar (dereferencing the `Box`), you'd see the correct integer type.

Unfortunately, Rust playground doesn't work at the moment, so I can't update the code there
</blockquote></details>

It looks like `as_ref()` yield type `dyn Any` but not the type behind `Any`?

**Review by schreter — APPROVED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1206)* status: all files reviewed, 1 unresolved discussion (waiting on @drmingdrmer)

___
*[`openraft/src/core/sm/worker.rs` line 146 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2oa3R939S5bECfB19_:-O2uedFUCSsCYZlw9uN0:bkxi2vx) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/core/sm/worker.rs#L146)):*
<details><summary><i>Previously, drmingdrmer (张炎泼) wrote…</i></summary><blockquote>

It looks like `as_ref()` yield type `dyn Any` but not the type behind `Any`?


</blockquote></details>

That's correct. But the implementation of `Foo` on this type (`i32`) is that returning `"i32"`. Note that the trait implementation is generated for the *concrete* type. The call on `dyn Trait` then dispatches the call to the generated function hidden behind object type `dyn Trait`. Also, note that the `Box` is then not on `dyn Any` but rather on `dyn Foo`, since that's the type you require. But, 

See the fixed example at https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9c305a6e746a978c99fc0e845b5d3cca, it will print `i32`.

**Review by drmingdrmer — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1206)* status: all files reviewed, 1 unresolved discussion

___
*[`openraft/src/core/sm/worker.rs` line 146 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2oa3R939S5bECfB19_:-O2wN-1FAQ4EU3l-DU-j:b-wo06wq) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/core/sm/worker.rs#L146)):*
<details><summary><i>Previously, schreter wrote…</i></summary><blockquote>

That's correct. But the implementation of `Foo` on this type (`i32`) is that returning `"i32"`. Note that the trait implementation is generated for the *concrete* type. The call on `dyn Trait` then dispatches the call to the generated function hidden behind object type `dyn Trait`. Also, note that the `Box` is then not on `dyn Any` but rather on `dyn Foo`, since that's the type you require. But, 

See the fixed example at https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9c305a6e746a978c99fc0e845b5d3cca, it will print `i32`.
</blockquote></details>

I see. But the problem here is that I can not use a customized type like using `Foo`. 
The type here is `FnOnce(&mut SM)`, there is a type parameter `SM` that must be specified.
But the channel for sending this command(`enum RaftMsg<C>`) does not have this type parameter.

**Review by schreter — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1206)* status: all files reviewed, 1 unresolved discussion

___
*[`openraft/src/core/sm/worker.rs` line 146 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1206#-O2oa3R939S5bECfB19_:-O2xWwGw6JkTqt8yG5-k:bks30oo) ([raw file](https://github.com/datafuselabs/openraft/blob/42b538181a2f6770647eaad5102031c15e9ba4e4/openraft/src/core/sm/worker.rs#L146)):*
<details><summary><i>Previously, drmingdrmer (张炎泼) wrote…</i></summary><blockquote>

I see. But the problem here is that I can not use a customized type like using `Foo`. 
The type here is `FnOnce(&mut SM)`, there is a type parameter `SM` that must be specified.
But the channel for sending this command(`enum RaftMsg<C>`) does not have this type parameter.
</blockquote></details>

Sure, but I was suggesting to add something like this:

```rust
trait CallSM: OptionalSend {
    fn sm_type(&self) -> &'static str;
    fn call(self, sm: &mut dyn Any) -> Box<dyn Future<Output = Result<(), ...>>>;
}

trait InnerFunc<SM>: FnOnce<&mut SM> + OptionalSend {}
impl<SM, F: FnOnce<&mut SM> + OptionalSend> InnerFunc<SM> for F {}

pub struct Call<SM, F: InnerFunc<SM>>(F);

impl<SM, F: InnerFunc<SM>> CallSM for Call<SM, F> {
    fn sm_type(&self) -> &'static str { std::any::type_name::<SM>() }
    fn call(self, sm: &mut dyn Any) -> Result<Box<dyn Future<Output = ()>>, ...> {
        if let Some(sm) = sm.downcast_ref::<SM>() {
            Ok((self.0)(sm))
        } else {
            Err(...)
        }
    }
}
```

Then, pack the function into `Call` and use `Box<dyn CallSM>` to transfer it. At the receiving side, store the type, then call the function and if it fails, use the type in the error message.

File: openraft/src/raft/mod.rs [hunk 1]
```diff
@@ -780,10 +780,95 @@ where C: RaftTypeConfig
...
+    /// The request functor will be called with a mutable reference to the state machine.
+    /// The functor returns a [`Future`] because state machine methods are `async`.
+    ///
+    /// If the API channel is already closed (Raft is in shutdown), then the request functor is
+    /// destroyed right away and not called at all.
+    ///
+    /// If the input `SM` is different from the one in `RaftCore`, it just silently ignores it.
+    pub fn external_state_machine_request<F, SM>(&self, req: F)
+    where
+        SM: 'static,
```
Comment (SteveLauC) [inline openraft/src/raft/mod.rs:854]: Noticed that `SM` does not have a `RaftStateMachine<C>` trait bound, is this something by design for greater flexibility? 

```rust
SM: 'static + RaftStateMachine<C>,
```

Though `SM` can indeed be anything as the underlying state machine worker would ignore incorrect type:

```rust
                    } else {
                        tracing::warn!(
                            "User-defined SM function uses incorrect state machine type, expected: {}, got: {}",
                            std::any::type_name::<SM>(),
                            input_sm_type
                        );
                    };
```

Comment (drmingdrmer) [reply openraft/src/raft/mod.rs:854]: It must be a mistake 🤔

Comment (SteveLauC) [reply openraft/src/raft/mod.rs:854]: then let's fix it:)

**Pull request merged**